bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerHandle.java

```diff
@@ -1808,77 +1745,31 @@ void sendAddSuccessCallbacks() {
-                        metadata.getEnsembleSize(),
-                        metadata.getWriteQuorumSize(),
-                        metadata.getAckQuorumSize(),
-                        metadata.getCustomMetadata(),
-                        newEnsemble,
-                        idx,
-                        new HashSet<BookieSocketAddress>(failedBookies.values()));
-                    newEnsemble.set(idx, newBookie);
-                    replacedBookies.add(idx);
-                } catch (BKException.BKNotEnoughBookiesException e) {
-                    // if there is no bookie replaced, we throw not enough bookie exception
-                    if (replacedBookies.size() <= 0) {
-                        throw e;
-                    } else {
-                        break;
-                    }
-                }
-            }
-            if (LOG.isDebugEnabled()) {
-                LOG.debug("[EnsembleChange-L{}-{}] : changing ensemble from: {} to: {} starting at entry: {},"
-                    + " failed bookies: {}, replaced bookies: {}",
-                        ledgerId, ensembleChangeIdx, getCurrentEnsemble(), newEnsemble,
-                        (getLastAddConfirmed() + 1), failedBookies, replacedBookies);
-            }
-            metadata.addEnsemble(newEnsembleStartEntry, newEnsemble);
+    void notifyWriteFailed(int index, BookieSocketAddress addr) {
+        synchronized (metadataLock) {
+            delayedWriteFailedBookies.put(index, addr);
         }
-        return new EnsembleInfo(newEnsemble, failedBookies, replacedBookies);
     }
 
-    void handleDelayedWriteBookieFailure() {
-        final Map<Integer, BookieSocketAddress> copyDelayedWriteFailedBookies =
-                new HashMap<Integer, BookieSocketAddress>(delayedWriteFailedBookies);
-        delayedWriteFailedBookies.clear();
+    void maybeHandleDelayedWriteBookieFailure() {
+        synchronized (metadataLock) {
+            if (delayedWriteFailedBookies.isEmpty()) {
+                return;
+            }
+            Map<Integer, BookieSocketAddress> toReplace = new HashMap<>();
+            toReplace.putAll(delayedWriteFailedBookies);
+            delayedWriteFailedBookies.clear();
 
-        // Original intent of this change is to do a best-effort ensemble change.
-        // But this is not possible until the local metadata is completely immutable.
-        // Until the feature "Make LedgerMetadata Immutable #610" Is complete we will use
-        // handleBookieFailure() to handle delayed writes as regular bookie failures.
-        handleBookieFailure(copyDelayedWriteFailedBookies);
+            // Original intent of this change is to do a best-effort ensemble change.
+            // But this is not possible until the local metadata is completely immutable.
+            // Until the feature "Make LedgerMetadata Immutable #610" Is complete we will use
+            // handleBookieFailure() to handle delayed writes as regular bookie failures.
```

Comment:
This comment probably needs to be rewritten to explain precisely why we need to block acks and what would need to change.

Reply:
Yep.  That comment seems to imply that this refactor would fix it, so at least that bit should be updated.

Reply:
>  The whole deferred handle failure needs to change, 

@ivankelly can you flush out your thoughts on this?
I agree with modifying comments as per the discussion above.